include/memory_pool/shared_multi_heap.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2021 Carlo Caione, <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_MEMORY_POOL_SHARED_MULTI_HEAP_H_
+#define ZEPHYR_INCLUDE_MEMORY_POOL_SHARED_MULTI_HEAP_H_
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Shared multi-heap memory pool interface
+ * @defgroup shared_multi_heap Shared multi-heap memory pool interface
+ * @ingroup multi_heap
+ * @{
+ *
+ * The shared multi-heap memory pool manager uses the multi-heap allocator to
+ * manage a set of reserved memory regions with different capabilities /
+ * attributes (cacheable, non-cacheable, etc...) defined in the DT.
+ *
+ * The user can request allocation from the shared pool specifying the
+ * capability / attribute of interest for the memory (cacheable / non-cacheable
+ * memory, etc...)
+ *
+ */
+
+/**
+ * @brief Memory region attributes / capabilities
+ *
+ * !!! This list needs to be kept in sync with shared-multi-heap-pool.yaml !!!
+ */
+enum smh_reg_attr {
+	SMH_REG_ATTR_CACHEABLE,
+	SMH_REG_ATTR_NON_CACHEABLE,
+	SMH_REG_ATTR_NUM,
+};
+
+/**
+ * @brief SMH region struct
+ *
+ * This struct is carrying information about the memory region to be added in
+ * the multi-heap pool. This is filled by the manager with the information
+ * coming from the reserved memory children nodes in the DT.
+ */
+struct smh_region {
+	enum smh_reg_attr attr;
+	uintptr_t addr;
+	size_t size;
+};
+
+/**
+ * @brief Region init function
+ *
+ * This is a user-provided function whose responsibility is to setup or
+ * initialize the memory region passed in input before this is added to the
+ * heap pool by the shared multi-heap memory pool manager. This function can be
+ * used by architectures using MMU / MPU that must correctly map the region
+ * before this is considered valid and accessible.
+ *
+ * @param reg Pointer to the SMH region structure.
+ * @param v_addr Virtual address obtained after mapping. For non-MMU
+ *               architectures this value is the physical address of the
+ *               region.
+ * @param size Size of the region after mapping.
+ *
+ * @return True if the region is ready to be added to the heap pool.
+ *         False if the region must be skipped.
+ */
+typedef bool (*smh_init_reg_fn_t)(struct smh_region *reg, uint8_t **v_addr, size_t *size);
+
+
+/**
+ * @brief Init the pool
+ *
+ * Initialize the shared multi-heap memory pool and hook-up the region init
+ * function.
+ *
+ * @param smh_init_reg_fn The function pointer to the region init function. Can
+ *                        be NULL for non-MPU / non-MMU architectures.
+ */
+int smh_pool_init(smh_init_reg_fn_t smh_init_reg_fn);
```

Comment:
If it's a single global device instantiated out of devicetree, why bother exposing the initialization hook in the public API?  Seems like it could just live as a SYS_INIT() callback internally.

Reply:
But how do you pass the `smh_init_reg_fn` function to the framework in that case? `smh_init_reg_fn` is a user-provided function that is used as callback for the memory regions initialization.